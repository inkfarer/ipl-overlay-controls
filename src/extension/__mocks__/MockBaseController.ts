import { MessageInputMap, MessageResultMap } from '../../types/messages/MessageMaps';
import { BaseController } from '../controllers/BaseController';

export let controllerListeners:
    {[key in keyof MessageInputMap]?: (data?: unknown) => MessageResultMap[key] | Promise<MessageResultMap[key]>} = {};

beforeEach(() => {
    jest.spyOn(BaseController.prototype, 'listen').mockImplementation(
        <key extends keyof MessageInputMap>
        (name: key, callback: (data?: unknown) => any) => {
            controllerListeners[name] = callback;
        });

    controllerListeners = {};
});
