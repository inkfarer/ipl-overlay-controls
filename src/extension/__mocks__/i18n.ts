import i18next from 'i18next';

beforeAll(async () => {
    await i18next.init({
        lng: 'cimode',
        appendNamespaceToCIMode: true
    });
});
