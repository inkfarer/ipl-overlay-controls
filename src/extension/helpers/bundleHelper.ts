import { existsSync, readdirSync, readFileSync } from 'fs';
import * as nodecgContext from './nodecg';
import { NodecgBundle } from '../../types/nodecgBundle';
import { GameVersion } from '../../types/enums/gameVersion';
import i18next from 'i18next';

const nodecg = nodecgContext.get();
const bundlesDir = `${process.cwd()}/bundles`;
const allGameVersions = Object.keys(GameVersion);

export const dependentBundles: Array<NodecgBundle> = readdirSync(bundlesDir, { withFileTypes: true })
    .filter(dir => dir.isDirectory() && dir.name !== nodecg.bundleName)
    .map(bundle => {
        const bundlePackagePath = `${bundlesDir}/${bundle.name}/package.json`;
        if (existsSync(bundlePackagePath)) {
            const parsedPackage = JSON.parse(readFileSync(bundlePackagePath, 'utf-8'));
            const gameVersions = normalizeGameVersions(parsedPackage.compatibleGameVersions);
            if (gameVersions.length <= 0) {
                nodecg.log.warn(i18next.t('bundleHelper.gameVersionParsingFailedWarning', { bundleName: bundle.name }));
            }

            if ('compatibleDashboardVersion' in parsedPackage) {
                return ({
                    name: bundle.name,
                    compatibleDashboardVersion: parsedPackage.compatibleDashboardVersion,
                    compatibleGameVersions: gameVersions ?? []
                });
            }
        }

        return null;
    })
    .filter(Boolean);

function normalizeGameVersions(versions?: Array<string> | null): Array<string> {
    if (!Array.isArray(versions) || versions.length <= 0) {
        return [ GameVersion.SPLATOON_2 ];
    }

    return versions.filter(version => allGameVersions.includes(version));
}
