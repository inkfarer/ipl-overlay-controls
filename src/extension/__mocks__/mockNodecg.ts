import express from 'express';
import last from 'lodash/last';
import { Configschema } from '../../types/schemas';
import cloneDeep from 'lodash/cloneDeep';

export type ReplicantChangeHandler = (newValue?: unknown, oldValue?: unknown) => void;
type OnFunction = (event: string, handler: ReplicantChangeHandler) => void;

const defaultBundleConfig = Object.freeze({
    lastfm: {
        apiKey: 'lastfmkey123',
        secret: 'lastfmsecret456'
    },
    smashgg: {
        apiKey: 'smashggkey789'
    },
    radia: {
        url: 'radia://url',
        socketUrl: 'ws://radia.url',
        authentication: 'radia-auth-12345'
    }
});

export let replicants: {[key: string]: unknown} = {};
export const replicantChangeListeners: {[key: string]: ReplicantChangeHandler} = {};
export const requestHandlers: {[type in 'POST' | 'GET']: {[path: string]: express.RequestHandler}}
    = { POST: {}, GET: {} };
export const mockBundleConfig: Configschema = cloneDeep(defaultBundleConfig);
export const messageListeners: {[key: string]: (message?: unknown, cb?: () => void) => void} = {};
export const mockMount = jest.fn();
export const mockNodecgLog = {
    warn: jest.fn(),
    info: jest.fn(),
    error: jest.fn()
};
export const mockSendMessage = jest.fn();

beforeEach(() => {
    replicants = {};
    Object.assign(mockBundleConfig, cloneDeep(defaultBundleConfig));
});

require('../helpers/nodecg').set({
    Replicant(name: string) {
        const replicantValue: { value: unknown, on: OnFunction } = {
            get value() {
                return replicants[name];
            },
            set value(newValue: unknown) {
                replicants[name] = newValue;
            },
            on: (event: string, handler: ReplicantChangeHandler) => {
                replicantChangeListeners[name] = handler;
            }
        };
        return replicantValue;
    },
    listenFor: (messageName: string, handler: () => void) => {
        messageListeners[messageName] = handler as (message: unknown, cb?: () => void) => void;
    },
    Router: () => ({
        post(path: string, ...handlers: express.RequestHandler[]) {
            requestHandlers['POST'][path] = last(handlers);
        },
        get(path: string, ...handlers: express.RequestHandler[]) {
            requestHandlers['GET'][path] = last(handlers);
        }
    }),
    mount: mockMount,
    log: mockNodecgLog,
    bundleConfig: mockBundleConfig,
    bundleName: 'ipl-overlay-controls',
    sendMessage: mockSendMessage
});
