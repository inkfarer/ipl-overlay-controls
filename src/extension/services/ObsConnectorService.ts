import type NodeCG from '@nodecg/types';
import { ObsCredentials, ObsData } from 'schemas';
import OBSWebSocket, { EventTypes } from 'obs-websocket-js';
import { ObsStatus } from 'types/enums/ObsStatus';
import { isBlank } from '../../helpers/stringHelper';

// Authentication failed, Unsupported protocol version, Session invalidated
const SOCKET_CLOSURE_CODES_FORBIDDING_RECONNECTION = [4009, 4010, 4011];

export class ObsConnectorService {
    private readonly nodecg: NodeCG.ServerAPI;
    private obsData: NodeCG.ServerReplicant<ObsData>;
    private obsCredentials: NodeCG.ServerReplicant<ObsCredentials>;
    private socket: OBSWebSocket;
    private reconnectionInterval: NodeJS.Timeout;
    private reconnectionCount: number;

    constructor(nodecg: NodeCG.ServerAPI) {
        this.nodecg = nodecg;
        this.obsData = nodecg.Replicant('obsData');
        this.obsCredentials = nodecg.Replicant('obsCredentials');
        this.socket = new OBSWebSocket();
        this.reconnectionCount = 0;

        this.socket.on('ConnectionClosed', e => this.handleClosure(e));
        this.socket.on('ConnectionOpened', () => this.handleOpening());
        this.socket.on('SceneListChanged', e => this.handleSceneListChange(e));
        this.socket.on('CurrentProgramSceneChanged', e => this.handleProgramSceneChange(e));

        if (this.obsData.value.enabled) {
            this.connect().catch(() => {
                // ignore
            });
        }
    }

    private handleClosure(event: EventTypes['ConnectionClosed']): void {
        if (this.obsData.value.status === ObsStatus.CONNECTED) {
            if (event.code !== 1000) {
                this.nodecg.log.error('OBS websocket closed with message:', event.message);
            }
            this.obsData.value.status = ObsStatus.NOT_CONNECTED;
            if (this.obsData.value.enabled) {
                this.startReconnecting(event.code);
            }
        }
    }

    private handleOpening(): void {
        this.nodecg.log.info('OBS websocket is open.');
        this.obsData.value.status = ObsStatus.CONNECTED;
        this.stopReconnecting();
    }

    private handleSceneListChange(event: EventTypes['SceneListChanged']): void {
        this.updateScenes(event.scenes.map(scene => String(scene.sceneName)));
    }

    private handleProgramSceneChange(event: EventTypes['CurrentProgramSceneChanged']): void {
        this.obsData.value.currentScene = event.sceneName;
    }

    async connect(reconnectOnFailure = true): Promise<void> {
        await this.socket.disconnect();
        this.obsData.value.status = ObsStatus.CONNECTING;

        try {
            // todo: if address does not start with http(s), add it manually (client side?)
            await this.socket.connect(
                this.obsCredentials.value.address,
                isBlank(this.obsCredentials.value.password) ? undefined : this.obsCredentials.value.password);
        } catch (e) {
            this.obsData.value.status = ObsStatus.NOT_CONNECTED;
            if (reconnectOnFailure) {
                this.startReconnecting(e.code);
            }
            throw new Error(`Failed to connect to OBS: ${e.message ?? String(e)}`);
        }

        await this.loadSceneList();
    }

    async disconnect(): Promise<void> {
        this.stopReconnecting();
        await this.socket.disconnect();
    }

    private async loadSceneList(): Promise<void> {
        const scenes = await this.socket.call('GetSceneList');

        this.obsData.value.currentScene = scenes.currentProgramSceneName;
        this.updateScenes(scenes.scenes.map(scene => String(scene.sceneName)));
    }

    startReconnecting(socketClosureCode?: number): void {
        if (SOCKET_CLOSURE_CODES_FORBIDDING_RECONNECTION.includes(socketClosureCode)) return;

        this.stopReconnecting();
        this.reconnectionInterval = setInterval(() => {
            this.reconnectionCount++;
            this.nodecg.log.info(`Attempting to reconnect to OBS... (Attempt ${this.reconnectionCount})`);
            this.connect(false).catch(() => {
                // ignore
            });
        }, 10000);
    }

    stopReconnecting(): void {
        clearInterval(this.reconnectionInterval);
        this.reconnectionInterval = null;
        this.reconnectionCount = 0;
    }

    private updateScenes(scenes: string[]): void {
        // OBS does not allow you to have no scenes.
        if (scenes.length <= 0) {
            this.nodecg.log.error('Received scene list with no scenes.');
            return;
        }

        const obsDataUpdates: Record<string, unknown> = {};
        if (!scenes.includes(this.obsData.value.gameplayScene)) {
            obsDataUpdates.gameplayScene = scenes[0];
        }
        if (!scenes.includes(this.obsData.value.intermissionScene)) {
            obsDataUpdates.intermissionScene = scenes[scenes.length === 1 ? 0 : 1];
        }
        obsDataUpdates.scenes = scenes;
        this.obsData.value = {
            ...this.obsData.value,
            ...obsDataUpdates
        };

        if (obsDataUpdates.gameplayScene || obsDataUpdates.intermissionScene) {
            this.nodecg.sendMessage('obsSceneConfigurationChangedAfterUpdate');
        }
    }

    setCurrentScene(scene: string): Promise<void> {
        return this.socket.call('SetCurrentProgramScene', { sceneName: scene });
    }
}
