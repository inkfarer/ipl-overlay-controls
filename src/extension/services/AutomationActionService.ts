/*
 * ipl-overlay-controls contains a "Game automation actions" module.
 * Each action has a list of pre-defined "Tasks" that it completes in a pre-defined order, each one at its own time.
 */

import type NodeCG from '@nodecg/types';
import { GameAutomationData, ObsData, RuntimeConfig, ScoreboardData } from 'schemas';
import { GameVersion } from 'types/enums/gameVersion';
import { GameAutomationAction } from 'types/enums/GameAutomationAction';
import { switchToNextColor } from '../replicants/activeRound';
import { ObsConnectorService } from './ObsConnectorService';
import * as util from 'util';

interface AutomationActionTask {
    timeout: number
    name: string
    action: () => unknown
}

interface GameStartTimings {
    showScoreboard: number
    showCasters: number
}

interface GameEndTimings {
    hideScoreboard: number
    changeScene: number
}

const startTimings: Record<GameVersion, GameStartTimings> = {
    [GameVersion.SPLATOON_2]: {
        showScoreboard: 11500,
        showCasters: 12000
    },
    [GameVersion.SPLATOON_3]: {
        showScoreboard: 11500,
        showCasters: 12000
    }
};

const endTimings: Record<GameVersion, GameEndTimings> = {
    [GameVersion.SPLATOON_2]: {
        hideScoreboard: 3000,
        changeScene: 7500
    },
    [GameVersion.SPLATOON_3]: {
        hideScoreboard: 3000,
        changeScene: 7500
    }
};

export class AutomationActionService {
    private nodecg: NodeCG.ServerAPI;
    private obsData: NodeCG.ServerReplicant<ObsData>;
    private scoreboardData: NodeCG.ServerReplicant<ScoreboardData>;
    private runtimeConfig: NodeCG.ServerReplicant<RuntimeConfig>;
    private gameAutomationData: NodeCG.ServerReplicant<GameAutomationData>;
    private readonly obsConnectorService: ObsConnectorService;
    private automationTasks: Array<AutomationActionTask> | null;
    private nextAutomationTaskTimeout: NodeJS.Timeout;

    constructor(nodecg: NodeCG.ServerAPI, obsConnectorService: ObsConnectorService) {
        this.nodecg = nodecg;
        this.obsData = nodecg.Replicant<ObsData>('obsData');
        this.scoreboardData = nodecg.Replicant<ScoreboardData>('scoreboardData');
        this.runtimeConfig = nodecg.Replicant<RuntimeConfig>('runtimeConfig');
        this.gameAutomationData = nodecg.Replicant<GameAutomationData>('gameAutomationData');
        this.obsConnectorService = obsConnectorService;
        this.automationTasks = null;
    }

    private getAutomationTasks(action: GameAutomationAction): Array<AutomationActionTask> {
        const gameVersion = this.runtimeConfig.value.gameVersion;
        switch (action) {
            case GameAutomationAction.START_GAME:
                return [
                    {
                        timeout: 1000,
                        name: 'changeScene',
                        action: async () => {
                            switchToNextColor();
                            await this.obsConnectorService.setCurrentScene(this.obsData.value.gameplayScene);
                        }
                    },
                    {
                        timeout: startTimings[gameVersion].showScoreboard,
                        name: 'showScoreboard',
                        action: () => {
                            this.scoreboardData.value.isVisible = true;
                        }
                    },
                    {
                        timeout: startTimings[gameVersion].showCasters,
                        name: 'showCasters',
                        action: () => {
                            this.nodecg.sendMessage('mainShowCasters');
                        }
                    }
                ];
            case GameAutomationAction.END_GAME:
                return [
                    {
                        timeout: endTimings[gameVersion].hideScoreboard,
                        name: 'hideScoreboard',
                        action: () => {
                            this.scoreboardData.value.isVisible = false;
                        }
                    },
                    {
                        timeout: endTimings[gameVersion].changeScene,
                        name: 'changeScene',
                        action: async () => {
                            await this.obsConnectorService.setCurrentScene(this.obsData.value.intermissionScene);
                        }
                    }
                ];
            default:
                throw new Error(`Unknown GameAutomationTask value '${action}'`);
        }
    }

    startAutomationAction(action: GameAutomationAction): void {
        if (this.gameAutomationData.value.actionInProgress !== GameAutomationAction.NONE) {
            throw new Error('An action is already in progress.');
        }

        this.gameAutomationData.value.actionInProgress = action;
        this.automationTasks = this.getAutomationTasks(action);

        this.queueNextAutomationTask().catch(e => {
            this.nodecg.log.error('Failed to start automation task:', e.message ?? String(e));
        });
    }

    private async queueNextAutomationTask(): Promise<void> {
        let nextTaskIndex = (this.gameAutomationData.value.nextTaskForAction?.index ?? -1) + 1;
        let nextTask = this.automationTasks[nextTaskIndex];
        // Immediately execute tasks with no timeout without sending info about the tasks to clients
        while (nextTask?.timeout === 0) {
            await this.executeAutomationTask(nextTask);
            nextTaskIndex++;
            nextTask = this.automationTasks[nextTaskIndex];
        }
        if (!nextTask) {
            this.resetGameAutomationData();
        } else {
            this.gameAutomationData.value.nextTaskForAction = {
                index: nextTaskIndex,
                name: nextTask.name,
                executionTimeMillis: new Date().getTime() + nextTask.timeout
            };
            this.nextAutomationTaskTimeout = setTimeout(async () => {
                await this.completeTimedAutomationTask(nextTask);
            }, nextTask.timeout);
        }
    }

    private async completeTimedAutomationTask(task: AutomationActionTask): Promise<void> {
        clearTimeout(this.nextAutomationTaskTimeout);
        await this.executeAutomationTask(task);
        await this.queueNextAutomationTask();
    }

    private async executeAutomationTask(task: AutomationActionTask): Promise<void> {
        try {
            const result = task.action();
            if (util.types.isPromise(result)) {
                await result;
            }
        } catch (e) {
            this.nodecg.log.error('Encountered an error during automation task', e);
        }
    }

    async fastForwardToNextAutomationTask(): Promise<void> {
        if (this.gameAutomationData.value.actionInProgress === GameAutomationAction.NONE) {
            throw new Error('No action is in progress.');
        }

        await this.completeTimedAutomationTask(
            this.automationTasks[this.gameAutomationData.value.nextTaskForAction.index]);
    }

    cancelAutomationAction(): void {
        if (this.gameAutomationData.value.actionInProgress === GameAutomationAction.NONE) {
            throw new Error('No action is in progress.');
        }

        clearTimeout(this.nextAutomationTaskTimeout);
        this.resetGameAutomationData();
    }

    resetGameAutomationData(): void {
        this.gameAutomationData.value.actionInProgress = GameAutomationAction.NONE;
        this.gameAutomationData.value.nextTaskForAction = null;
        this.automationTasks = null;
    }
}
