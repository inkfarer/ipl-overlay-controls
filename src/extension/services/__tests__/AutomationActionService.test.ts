import { mock } from 'jest-mock-extended';

import * as ActiveRoundModule from '../../replicants/activeRound';
const mockActiveRoundModule = mock<typeof ActiveRoundModule>();
jest.mock('../../replicants/activeRound', () => mockActiveRoundModule);

import { cartesian } from '../../../helpers/ArrayHelper';
import { GameAutomationAction } from '../../../types/enums/GameAutomationAction';
import { GameVersion } from '../../../types/enums/gameVersion';
import { GameAutomationData, ScoreboardData } from '../../../types/schemas';
import { mockNodecg, replicants } from '../../__mocks__/mockNodecg';
import { AutomationActionService } from '../AutomationActionService';
import { ObsConnectorService } from '../ObsConnectorService';

describe('AutomationActionService', () => {
    let obsConnectorService: ObsConnectorService;

    beforeEach(() => {
        jest.clearAllMocks();
        jest.restoreAllMocks();
        jest.useFakeTimers();
        jest.spyOn(Date.prototype, 'getTime').mockReturnValue(10000);
        replicants.obsData = { gameplayScene: 'Gameplay Scene', intermissionScene: 'Break Scene' };
        replicants.scoreboardData = {};
        replicants.gameAutomationData = {
            actionInProgress: GameAutomationAction.NONE,
            nextTaskForAction: null
        };
        obsConnectorService = mock<ObsConnectorService>();
    });

    afterEach(() => {
        jest.clearAllTimers();
    });

    describe('getAutomationTasks', () => {
        beforeEach(() => {
            replicants.runtimeConfig = {
                gameVersion: GameVersion.SPLATOON_2
            };
        });

        it('returns expected actions for starting a game', async () => {
            replicants.obsData = {
                gameplayScene: 'gameplay scene'
            };
            replicants.scoreboardData = {
                isVisible: false
            };
            const service = new AutomationActionService(mockNodecg, obsConnectorService);

            const result = service['getAutomationTasks'](GameAutomationAction.START_GAME);

            expect(result.length).toEqual(3);

            await result[0].action();
            expect(mockActiveRoundModule.switchToNextColor).toHaveBeenCalled();
            expect(obsConnectorService.setCurrentScene).toHaveBeenCalledWith('gameplay scene');
            expect((replicants.scoreboardData as ScoreboardData).isVisible).toEqual(false);
            expect(mockNodecg.sendMessage).not.toHaveBeenCalled();

            result[1].action();
            expect((replicants.scoreboardData as ScoreboardData).isVisible).toEqual(true);
            expect(mockNodecg.sendMessage).not.toHaveBeenCalled();

            result[2].action();
            expect(mockNodecg.sendMessage).toHaveBeenCalledWith('mainShowCasters');
        });

        it('returns expected actions for ending a game', async () => {
            replicants.obsData = {
                intermissionScene: 'intermission scene'
            };
            replicants.scoreboardData = {
                isVisible: true
            };
            const service = new AutomationActionService(mockNodecg, obsConnectorService);

            const result = service['getAutomationTasks'](GameAutomationAction.END_GAME);

            expect(result.length).toEqual(2);

            result[0].action();
            expect((replicants.scoreboardData as ScoreboardData).isVisible).toEqual(false);
            expect(obsConnectorService.setCurrentScene).not.toHaveBeenCalled();

            await result[1].action();
            expect(obsConnectorService.setCurrentScene).toHaveBeenCalledWith('intermission scene');
        });

        it.each(
            cartesian(Object.values(GameVersion), Object.values(GameAutomationAction))
        )('has expected timings for game version %s and action %s', (version, action) => {
            if (action === GameAutomationAction.NONE) return;

            replicants.runtimeConfig = {
                gameVersion: version
            };

            const service = new AutomationActionService(mockNodecg, obsConnectorService);

            const result = service['getAutomationTasks'](action);

            expect(result).toMatchSnapshot();
        });

        it('throws an error on unknown automation tasks', () => {
            const service = new AutomationActionService(mockNodecg, obsConnectorService);

            // @ts-ignore
            expect(() => service['getAutomationTasks']('asdasdasdasd'))
                .toThrow(new Error('Unknown GameAutomationTask value \'asdasdasdasd\''));
        });
    });

    describe('startAutomationAction', () => {
        it('throws an error if an action is already in progress', () => {
            replicants.gameAutomationData = {
                actionInProgress: GameAutomationAction.START_GAME
            };
            const service = new AutomationActionService(mockNodecg, obsConnectorService);

            expect(() => service.startAutomationAction(GameAutomationAction.END_GAME))
                .toThrow(new Error('An action is already in progress.'));
        });

        it('updates automation data and starts the specified action', () => {
            replicants.gameAutomationData = {
                actionInProgress: GameAutomationAction.NONE
            };
            jest.spyOn(AutomationActionService.prototype as any, 'getAutomationTasks')
                .mockReturnValue('test automation tasks');
            jest.spyOn(AutomationActionService.prototype as any, 'queueNextAutomationTask').mockResolvedValue(null);
            const service = new AutomationActionService(mockNodecg, obsConnectorService);

            service.startAutomationAction(GameAutomationAction.START_GAME);

            expect((replicants.gameAutomationData as GameAutomationData).actionInProgress)
                .toEqual(GameAutomationAction.START_GAME);
            expect(service['getAutomationTasks']).toHaveBeenCalledWith(GameAutomationAction.START_GAME);
            expect(service['automationTasks']).toEqual('test automation tasks');
            expect(service['queueNextAutomationTask']).toHaveBeenCalled();
        });
    });

    describe('queueNextAutomationTask', () => {
        it('queues up the first automation task if no info about the next task is known', async () => {
            jest.spyOn(Date.prototype, 'getTime').mockReturnValue(99999);
            jest.spyOn(global, 'setTimeout');
            jest.spyOn(AutomationActionService.prototype as any, 'completeTimedAutomationTask').mockReturnValue(null);
            replicants.gameAutomationData = {
                nextTaskForAction: null
            };
            const service = new AutomationActionService(mockNodecg, obsConnectorService);
            const firstTask = {
                name: 'firstTask',
                timeout: 1001,
                action: jest.fn()
            };
            service['automationTasks'] = [
                firstTask,
                {
                    name: 'secondTask',
                    timeout: 1002,
                    action: jest.fn()
                }
            ];

            await service['queueNextAutomationTask']();

            expect((replicants.gameAutomationData as GameAutomationData).nextTaskForAction).toEqual({
                index: 0,
                name: 'firstTask',
                executionTimeMillis: 101000
            });
            expect(service['nextAutomationTaskTimeout']).not.toBeNull();
            expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 1001);
            const timeoutHandler = (setTimeout as unknown as jest.Mock).mock.calls[0][0];
            expect(service['completeTimedAutomationTask']).not.toHaveBeenCalled();

            await timeoutHandler();
            expect(service['completeTimedAutomationTask']).toHaveBeenCalledWith(firstTask);
        });

        it('queues up the automation action after the current one', async () => {
            jest.spyOn(Date.prototype, 'getTime').mockReturnValue(1111111);
            jest.spyOn(global, 'setTimeout');
            jest.spyOn(AutomationActionService.prototype as any, 'completeTimedAutomationTask').mockReturnValue(null);
            replicants.gameAutomationData = {
                nextTaskForAction: {
                    index: 0,
                    name: 'firstTask',
                    executionTimeMillis: 101000
                }
            };
            const service = new AutomationActionService(mockNodecg, obsConnectorService);
            const secondTask = {
                name: 'secondTask',
                timeout: 1002,
                action: jest.fn()
            };
            service['automationTasks'] = [
                {
                    name: 'firstTask',
                    timeout: 1001,
                    action: jest.fn()
                },
                secondTask
            ];

            await service['queueNextAutomationTask']();

            expect((replicants.gameAutomationData as GameAutomationData).nextTaskForAction).toEqual({
                index: 1,
                name: 'secondTask',
                executionTimeMillis: 1112113
            });
            expect(service['nextAutomationTaskTimeout']).not.toBeNull();
            expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 1002);
            const timeoutHandler = (setTimeout as unknown as jest.Mock).mock.calls[0][0];
            expect(service['completeTimedAutomationTask']).not.toHaveBeenCalled();

            await timeoutHandler();
            expect(service['completeTimedAutomationTask']).toHaveBeenCalledWith(secondTask);
        });

        it('clears automation data if the final task has been completed', async () => {
            jest.spyOn(global, 'setTimeout');
            jest.spyOn(AutomationActionService.prototype, 'resetGameAutomationData').mockReturnValue(null);
            replicants.gameAutomationData = {
                nextTaskForAction: {
                    index: 1,
                    name: 'secondTask',
                    executionTimeMillis: 39393939
                }
            };
            const service = new AutomationActionService(mockNodecg, obsConnectorService);
            service['automationTasks'] = [
                {
                    name: 'firstTask',
                    timeout: 1001,
                    action: jest.fn()
                },
                {
                    name: 'secondTask',
                    timeout: 1002,
                    action: jest.fn()
                }
            ];

            await service['queueNextAutomationTask']();

            expect(service.resetGameAutomationData).toHaveBeenCalled();
            expect(setTimeout).not.toHaveBeenCalled();
        });

        it('executes each task with no timeout before queueing the next one', async () => {
            jest.spyOn(Date.prototype, 'getTime').mockReturnValue(1111111);
            jest.spyOn(global, 'setTimeout');
            jest.spyOn(AutomationActionService.prototype as any, 'completeTimedAutomationTask').mockReturnValue(null);
            jest.spyOn(AutomationActionService.prototype as any, 'executeAutomationTask').mockReturnValue(null);
            replicants.gameAutomationData = {
                nextTaskForAction: {
                    index: 0,
                    name: 'firstTask',
                    executionTimeMillis: 101000
                }
            };
            const service = new AutomationActionService(mockNodecg, obsConnectorService);
            const secondTask = {
                name: 'secondTask',
                timeout: 0,
                action: jest.fn()
            };
            const thirdTask = {
                name: 'thirdTask',
                timeout: 0,
                action: jest.fn()
            };
            service['automationTasks'] = [
                {
                    name: 'firstTask',
                    timeout: 1001,
                    action: jest.fn()
                },
                secondTask,
                thirdTask,
                {
                    name: 'fourthTask',
                    timeout: 1004,
                    action: jest.fn()
                }
            ];

            await service['queueNextAutomationTask']();

            expect(service['executeAutomationTask']).toHaveBeenCalledTimes(2);
            expect(service['executeAutomationTask']).toHaveBeenCalledWith(secondTask);
            expect(service['executeAutomationTask']).toHaveBeenCalledWith(thirdTask);
            expect((replicants.gameAutomationData as GameAutomationData).nextTaskForAction).toEqual({
                index: 3,
                name: 'fourthTask',
                executionTimeMillis: 1112115
            });
            expect(service['nextAutomationTaskTimeout']).not.toBeNull();
            expect(setTimeout).toHaveBeenCalledWith(expect.any(Function), 1004);
        });
    });

    describe('completeTimedAutomationTask', () => {
        it('clears the timeout to complete the timed automation task, executes it and queues the next one', async () => {
            jest.spyOn(AutomationActionService.prototype as any, 'queueNextAutomationTask').mockReturnValue(null);
            jest.spyOn(AutomationActionService.prototype as any, 'executeAutomationTask').mockReturnValue(null);
            jest.spyOn(global, 'clearTimeout');
            const service = new AutomationActionService(mockNodecg, obsConnectorService);
            const timeout = setTimeout(jest.fn(), 1000);
            service['nextAutomationTaskTimeout'] = timeout;
            const task = {
                timeout: 999,
                name: 'task name',
                action: jest.fn()
            };

            await service['completeTimedAutomationTask'](task);

            expect(clearTimeout).toHaveBeenCalledWith(timeout);
            expect(service['executeAutomationTask']).toHaveBeenCalledWith(task);
            expect(service['queueNextAutomationTask']).toHaveBeenCalled();
        });
    });

    describe('executeAutomationTask', () => {
        it('completes the action of the given task', async () => {
            const action = jest.fn();
            const service = new AutomationActionService(mockNodecg, obsConnectorService);

            await service['executeAutomationTask']({
                timeout: 998,
                name: 'task name',
                action
            });

            expect(action).toHaveBeenCalled();
        });

        it('handles errors within the action', async () => {
            const action = jest.fn().mockRejectedValue('error!');
            const service = new AutomationActionService(mockNodecg, obsConnectorService);

            await expect(service['executeAutomationTask']({
                timeout: 998,
                name: 'task name',
                action
            })).resolves.toBeUndefined();

            expect(action).toHaveBeenCalled();
        });
    });

    describe('fastForwardToNextAutomationTask', () => {
        it('throws an error if no action is in progress', async () => {
            replicants.gameAutomationData = {
                actionInProgress: GameAutomationAction.NONE
            };
            const service = new AutomationActionService(mockNodecg, obsConnectorService);

            await expect(service.fastForwardToNextAutomationTask())
                .rejects.toThrow(new Error('No action is in progress.'));
        });

        it('completes the next automation task', async () => {
            jest.spyOn(AutomationActionService.prototype as any, 'completeTimedAutomationTask').mockReturnValue(null);
            replicants.gameAutomationData = {
                nextTaskForAction: {
                    index: 1,
                    name: 'secondTask',
                    executionTimeMillis: 39393939
                },
                actionInProgress: GameAutomationAction.START_GAME
            };
            const service = new AutomationActionService(mockNodecg, obsConnectorService);
            const secondTask = {
                name: 'secondTask',
                timeout: 1002,
                action: jest.fn()
            };
            service['automationTasks'] = [
                {
                    name: 'firstTask',
                    timeout: 1001,
                    action: jest.fn()
                },
                secondTask
            ];

            await service.fastForwardToNextAutomationTask();

            expect(service['completeTimedAutomationTask']).toHaveBeenCalledWith(secondTask);
        });
    });

    describe('resetGameAutomationData', () => {
        it('clears info about automation actions in progress', () => {
            replicants.gameAutomationData = {
                nextTaskForAction: {
                    index: 1,
                    name: 'secondTask',
                    executionTimeMillis: 39393939
                },
                actionInProgress: GameAutomationAction.START_GAME
            };
            const service = new AutomationActionService(mockNodecg, obsConnectorService);
            service['automationTasks'] = [
                {
                    name: 'firstTask',
                    timeout: 1001,
                    action: jest.fn()
                },
                {
                    name: 'secondTask',
                    timeout: 1002,
                    action: jest.fn()
                }
            ];

            service.resetGameAutomationData();

            expect(service['automationTasks']).toBeNull();
            expect(replicants.gameAutomationData).toEqual({
                actionInProgress: GameAutomationAction.NONE,
                nextTaskForAction: null
            });
        });
    });

    describe('cancelAutomationAction', () => {
        it('clears the timeout to execute the next automation action and resets automation data', () => {
            replicants.gameAutomationData = {
                actionInProgress: GameAutomationAction.START_GAME
            };
            jest.spyOn(AutomationActionService.prototype, 'resetGameAutomationData').mockReturnValue(null);
            jest.spyOn(global, 'clearTimeout');
            const service = new AutomationActionService(mockNodecg, obsConnectorService);
            const timeout = setTimeout(jest.fn(), 1000);
            service['nextAutomationTaskTimeout'] = timeout;

            service.cancelAutomationAction();

            expect(service.resetGameAutomationData).toHaveBeenCalled();
            expect(clearTimeout).toHaveBeenCalledWith(timeout);
        });

        it('throws an error if no action is in progress', () => {
            replicants.gameAutomationData = {
                actionInProgress: GameAutomationAction.NONE
            };
            jest.spyOn(AutomationActionService.prototype, 'resetGameAutomationData');
            jest.spyOn(global, 'clearTimeout');
            const service = new AutomationActionService(mockNodecg, obsConnectorService);

            expect(() => service.cancelAutomationAction()).toThrow(new Error('No action is in progress.'));

            expect(service.resetGameAutomationData).not.toHaveBeenCalled();
            expect(clearTimeout).not.toHaveBeenCalled();
        });
    });
});
