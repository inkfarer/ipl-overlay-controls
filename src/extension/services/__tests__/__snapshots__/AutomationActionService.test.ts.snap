// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutomationActionService getAutomationTasks has expected timings for game version SPLATOON_2 and action END_GAME 1`] = `
[
  {
    "action": [Function],
    "name": "hideScoreboard",
    "timeout": 3000,
  },
  {
    "action": [Function],
    "name": "changeScene",
    "timeout": 7500,
  },
]
`;

exports[`AutomationActionService getAutomationTasks has expected timings for game version SPLATOON_2 and action START_GAME 1`] = `
[
  {
    "action": [Function],
    "name": "changeScene",
    "timeout": 1000,
  },
  {
    "action": [Function],
    "name": "showScoreboard",
    "timeout": 11500,
  },
  {
    "action": [Function],
    "name": "showCasters",
    "timeout": 12000,
  },
]
`;

exports[`AutomationActionService getAutomationTasks has expected timings for game version SPLATOON_3 and action END_GAME 1`] = `
[
  {
    "action": [Function],
    "name": "hideScoreboard",
    "timeout": 3000,
  },
  {
    "action": [Function],
    "name": "changeScene",
    "timeout": 7500,
  },
]
`;

exports[`AutomationActionService getAutomationTasks has expected timings for game version SPLATOON_3 and action START_GAME 1`] = `
[
  {
    "action": [Function],
    "name": "changeScene",
    "timeout": 1000,
  },
  {
    "action": [Function],
    "name": "showScoreboard",
    "timeout": 11500,
  },
  {
    "action": [Function],
    "name": "showCasters",
    "timeout": 12000,
  },
]
`;
